### ISSUE TITLE
docker rmi "image@digest" should not "untag" digests

### ISSUE THREAD

**Issue by thaJeztah** [docker/docker#24688]
(relates to https://github.com/docker/docker/pull/23996, https://github.com/docker/docker/pull/23649)

While working with 1.12-rc4, I noticed some UX inconsistencies;

Pulling an image looks good (much improved);

```bash
docker pull hello-world
Using default tag: latest
latest: Pulling from library/hello-world
Digest: sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9
Status: Downloaded newer image for hello-world:latest
```

Looking good:

```bash
docker images --digests --no-trunc hello-world
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID                                                                  CREATED             SIZE
hello-world         latest              sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9   sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB
```

However, when trying to remove an image, using the digest;

```bash
docker rmi hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9
Untagged: hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9
```

It's only removing the *digest*. This is confusing:

```bash
REPOSITORY          TAG                 DIGEST              IMAGE ID                                                                  CREATED             SIZE
hello-world         latest              <none>              sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB
```

**I don't see any practical use for just removing the digest.**

Using the image-id, on the other hand:

```bash
docker rmi sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc
Untagged: hello-world:latest
Untagged: hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9
Deleted: sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc
Deleted: sha256:a02596fdd012f22b03af6ad7d11fa590c57507558357b079c3e8cebceb4262d7
```

Actually deletes the image.

Similar, in a situation where the image is tagged under two names;

```bash
docker pull hello-world
….
docker tag hello-world hello-world:2
```

```bash
docker images --digests --no-trunc hello-world
REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID                                                                  CREATED             SIZE
hello-world         2                   sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9   sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB
hello-world         latest              sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9   sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB

docker rmi hello-world@sha256:0256e8a36e2070f7bf2d0b0763dbabdd67798512411de4cdcf9431a1feb60fd9
...

REPOSITORY          TAG                 DIGEST              IMAGE ID                                                                  CREATED             SIZE
hello-world         2                   <none>              sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB
hello-world         latest              <none>              sha256:c54a2cc56cbb2f04003c1cd4507e118af7c0d340fe7e2720f70976c4b75237dc   13 days ago         1.848 kB
```

Also, just the digest is removed

**Expected result**

I expect the digest to work similar to using the image-id, so, *delete* the image if possible, or produce an error if not. But *don't* remove just the digest.

**Comment by thaJeztah** [comment]
ping @dmcgowan @stevvooe @dnephin 

I added this to the 1.12 milestone, as this was improved in 1.12

**Comment by dnephin** [comment]
I think this is correct. The digest is working the same way as a tag.

**Comment by thaJeztah** [comment]
I think the difference is that I cannot "re-tag" an image with a different digest (unless the content changes), so from a user perspective, its an immutable identifier. Removing that identifier feels kinda "weird", and I'm not sure why we would do that.

**Comment by dnephin** [comment]
You can always re-pull to get the digest again.

This behaviour hasn't changed since 1.11:

```
$ docker version
...
 Version:      1.11.0-rc4
$ docker pull redis
...
Digest: sha256:b50f15d427aea5b579f9bf972ab82ff8c1c47bffc0481b225c6a714095a9ec34
Status: Downloaded newer image for redis:latest

$ docker pull redis@sha256:b50f15d427aea5b579f9bf972ab8
...
$ docker rmi redis@sha256:b50f15d427aea5b579f9bf972ab82ff8c1c47bffc0481b225c6a714095a9ec34
Untagged: redis@sha256:b50f15d427aea5b579f9bf972ab82ff8c1c47bffc0481b225c6a714095a9ec34


```

**Comment by MHBauer** [comment]
I agree that it's probably working as designed, but I would think that the digest is more part of the image, rather than a tag that can be removed.

I don't really want to see some special case in the code, but I am not sure what the right answer is.

The behavior is surprising in my opinion.

**Comment by dmcgowan** [comment]
> I expect the digest to work similar to using the image-id, so, delete the image if possible, or produce an error if not. But don't remove just the digest.

I think this is a reasonable expectation. We are trying to have the behavior of the digest be similar to image id, which up until this release has not been at all. Considering this is not a regression I would not give it a high priority, but I will work on a patch for it.

**Comment by aaronlehmann** [comment]
One thing to keep in mind is that an image may have multiple digests associated with it. It would be very surprising if untagging one of two digests removed a tag and deleted the image. Given that removing one of two digests probably shouldn't remove the image, I'm not sure we should introduce another special case where removing the single digest will remove the tag and delete the image.

I think before moving forward with this, we should agree on a complete matrix showing the expected behavior of deleting a tag, digest, or image ID when there are 0, 1, or 2 tags and 0, 1, or 2 digests.

**Comment by stevvooe** [comment]
What is going on here? Why are we "untagging" digests?

This is not even close to working as designed.

**Comment by dmcgowan** [comment]
@stevvooe "untagging" is misleading. We currently do treat digests as references which can be removed as references. As part of this change we should discuss what the behavior should be as Aaron mentioned. Unlike image ids, images can have multiple manifest digests. However we can always just treat them as image ids and remove in mass, the confusion comes if the "intended use" is to remove a digest. I think I agree that removing a digest form an image id does not need to and maybe should not be supported.

 Current `docker rmi` behavior

| Remove by| Tags | Digests | Removed references | Image deleted |
|---|---|---|---|---|
| tag | 1 | 0 | 1 tag | yes |
| tag | 1 | 1 | 1 tag, 1 digest | yes |
| tag | 1 | 2+ | 1 tag, all digests | yes |
| tag | 2+ | any | 1 tag | no |
| id | 0 | any | all digests | yes |
| id | 1 | 0 | 1 tag | yes|
| id | 1 | 1 | 1 tag, 1 digest | yes |
| id | 1 | 2+ | 1 tag, all digests | yes |
| id | 2+ | any | fails | no |
| digest | 0 | 1 | 1 digest | yes |
| digest | any | 2+ | 1 digest | no |
| digest | 1 | 1 | 1 digest | no |
| digest | 2+ | 1 | 1 digest| no |

The rows in question are the last 3, sounds like the desire behavior to make it more like image id would be.

| Remove by| Tags | Digests | Removed references | Image deleted |
|---|---|---|---|---|
| digest | 0 | any | all digests | yes |
| digest | 1 | any | 1 tag, all digests | yes |
| digest | 2+ | any | fails | no |

**Comment by thaJeztah** [comment]
I think @dmcgowan's proposed behavior looks right. One thing I'm a bit confused by is "multiple digests"; can you give an example when a single image has multiple digests (and how?), and what that would look like from a user perspective?

**Comment by dmcgowan** [comment]
Multiple digests can occur when pulling an image from multiple repositories which had layers with different digests. A different layer digest can occur anytime a layer ends up getting pushed, rather than mounted via cross repository mounted or skipped because it already exists. The formation of the tar should now be stable but we cannot guarantee the gzip algorithm will produce the exact same artifact digest, and therefore the same manifest digest. An example would be a user pulling down the image `ubuntu`, building an image from that image, and pushing it to a private repository. A new digest could be created during that push if any of the layers produce a different digest than what is pulled. We try to avoid it, but it is currently out of our control without having the original artifact to send to the private registry.

**Comment by thaJeztah** [comment]
Thanks for explaining! Wouldn't the image id be different in that case as well?

**Comment by aaronlehmann** [comment]
Negative. The image ID is not affected by compression of artifacts.

**Comment by dmcgowan** [comment]
I made a change to update the behavior and it feels right so far. The case where the image is tagged in multiple repositories is already handled by the multiple tag check (see below). In this case `-f` will force remove even across named repositories, which I think makes sense as well.
```
$ docker rmi something@sha256:661f2e4c9470e7f6238cebf603bcf5700c8b948894ac9e35f2cf6f63dcda723a
Error response from daemon: conflict: unable to delete a180b24e38ed (must be forced) - image is referenced in one or more repositories
$ docker rmi -f something@sha256:661f2e4c9470e7f6238cebf603bcf5700c8b948894ac9e35f2cf6f63dcda723a
Untagged: dmcgowan/something:latest
Untagged: something:latest
Untagged: something@sha256:661f2e4c9470e7f6238cebf603bcf5700c8b948894ac9e35f2cf6f63dcda723a
Deleted: sha256:a180b24e38eddf9e2133d07e9dc4c2ce7e74cb5adfe5bd34968a2b38f9685856
Deleted: sha256:da1ce37cb92a2d4704c21d44feb788b4b7312dc748ff086c093751b0a5e5434b
Deleted: sha256:5a56aca1dc3762f9080b27f4a38d08db26a59a0ea9b2c29cdcea7d5e00f7dbc6
Deleted: sha256:fbe47db404463ff45764edfd9e7a7d1b281e74284101cffd3b12bf6e5c33d9d7
```

Need to get some tests updated but have a work in progress branch here https://github.com/dmcgowan/docker/tree/rmi-manifest-digest-like-image-id, will cleanup and submit on wednesday when back from PTO.

**Comment by cpuguy83** [comment]
Is this really a 1.12 issue? Is this not present since 1.10?

**Comment by thaJeztah** [comment]
@cpuguy83 the behavior was changed for 1.12 to always store the digest; not sure if it's 1.12 specific, but would be good to have the behavior right

**Comment by thaJeztah** [comment]
In earlier versions, an image either had a digest, *or* a tag, but never both

**Comment by dnephin** [comment]
> In earlier versions, an image either had a digest, or a tag, but never both

Not true. It could have both, but a user would have had to explicitly pull by digest to get both, which was very rare.

> Is this really a 1.12 issue? 

No, it is not. I showed in https://github.com/docker/docker/issues/24688#issuecomment-232993537 that the issue is there in 1.11. It's just more prevalent with the changes in 1.12.

**Comment by thaJeztah** [comment]
@dnephin this is the behavior in 1.11;

```bash
docker pull alpine:latest
docker pull alpine@sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a

docker images --digests

REPOSITORY          TAG                 DIGEST                                                                    IMAGE ID            CREATED             SIZE
alpine              latest              <none>                                                                    4e38e38c8ce0        3 weeks ago         4.799 MB
alpine              <none>              sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a   4e38e38c8ce0        3 weeks ago         4.799 MB
```

**Comment by dnephin** [comment]
That's right. We changed the list output, but that is still a single image (`4e38e38c8ce0`) with both a tag and a digest.

**Comment by thaJeztah** [comment]
From a user perspective, they're two images, and in the 1.11 situation, "untagging" made sense, because the "image" I referred to disappears;

```
docker rmi alpine@sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a
Untagged: alpine@sha256:3dcdb92d7432d56604d4545cbd324b14e647b313626d99b889d0626de158f73a

docker images
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
alpine              latest              4e38e38c8ce0        3 weeks ago         4.799 MB
```

**Comment by runcom** [comment]
IMO this change should not be present in 1.12 as it didn't even make the feature freeze -  both referenced PRs were merged as part of 1.12-rc3. This should be deferred to 1.13.
